apistrap/flask.py

```diff
@@ -1,43 +1,178 @@
+        str: "string"
+    }
+
+    def __init__(self):
+        super().__init__()
+        self._app: Flask = None
+        self._specs_extracted = False
+
+    def init_app(self, app: Flask):
         """
-        Bind the extension to a Flask app
-        :param app: the Flask app to bind to
-        :param decorators: decorators that should be used to wrap the UI and specification views
+        Bind the extension to a Flask instance
+        :param app: the Flask instance
         """
+        self._app = app
+        blueprint = Blueprint("apistrap", __name__, template_folder=path.join(path.dirname(__file__), "templates"))
+
+        if self.spec_url is not None:
+            blueprint.route(self.spec_url, methods=["GET"])(self._get_spec)
+
+        if self.spec_url is not None and self.ui_url is not None:
+            blueprint.route(self.ui_url)(self._get_ui)
+            blueprint.route(self.ui_url + "/")(self._get_ui)
+
+        app.register_blueprint(blueprint)
+
         if self.use_default_error_handlers:
             app.register_error_handler(HTTPException, self.http_error_handler)
             app.register_error_handler(ApiClientError, self.error_handler)
             app.register_error_handler(ApiServerError, self.internal_error_handler)
             app.register_error_handler(Exception, self.internal_error_handler)
 
-        super().init_app(app, decorators)
+    def _is_bound(self) -> bool:
+        return self._app is not None
+
+    def _get_spec(self):
+        """
+        Serves the OpenAPI specification
+        """
+        self._extract_specs()
+        return jsonify(self.to_dict())
+
+    _get_spec.apistrap_ignore = True
+
+    def _get_ui(self):
+        """
+        Serves Swagger UI
+        """
+        return render_template("apidocs.html", apistrap=self)
+
+    _get_ui.apistrap_ignore = True
+
+    def _extract_specs(self):
+        """
+        Extract specification data from the Flask app and save it to the underlying Apispec object
+        """
+        if self._specs_extracted:
+            return
+
+        for rule in self._app.url_map.iter_rules():
+            # Skip Flask's internal static endpoint
+            if rule.endpoint == "static":
+                continue
+
+            handler = self._app.view_functions[rule.endpoint]
+
+            # Skip ignored endpoints
+            if getattr(handler, "apistrap_ignore", False):
+                continue
+
+            url = str(rule)
+            for arg in re.findall('(<([^<>]*:)?([^<>]*)>)', url):
+                url = url.replace(arg[0], '{%s}' % arg[2])
+
+            for method in rule.methods:
+                if method.lower() not in ["get", "post", "put", "delete", "patch"]:
+                    continue
+
+                self.spec.path(url, {
+                    method.lower(): self._extract_operation_specs(handler)
+                })
+
+        self._specs_extracted = True
+
+    def _extract_operation_specs(self, handler):
```

Comment:
This is awkwardly similar to the the implementation in `AioHTTPApistrap`...

Reply:
It is alright. I vote we wait for this to flesh out a bit more and then refactor it.

Reply:
Roger that